src/server/routes/oauth.ts

```diff
@@ -85,52 +86,111 @@ const redirectForGenericError = (_: Request, res: ResponseWithRequestState) => {
-
-	if (legacy_identity_id) {
-		// The legacy_identity_id claim is present, we can return the token set
-		return tokenSet;
-	}
+const fetchValidTokenSet = async ({
+	oktaId,
+	email,
+	openIdClient,
+	tokenSet,
+	refreshToken,
+	ip,
+	request_id,
+	maximumWaitTime = 15000,
+	waitTime = 1000,
+}: {
+	oktaId: string;
+	email?: string;
+	openIdClient: OpenIdClient;
+	tokenSet: TokenSet;
+	refreshToken: string;
+	ip: string;
+	request_id?: string;
+	maximumWaitTime?: number;
+	waitTime?: number;
+}): Promise<TokenSet | null> => {
+	const nonBlockingWait = async (
+		start: number,
+		maximumWaitTime: number,
+		waitTime: number,
+		refreshedTokenSet: TokenSet,
+		newRefreshToken: string,
+	): Promise<TokenSet | null> =>
+		// Non-blocking wait for the next iteration
+		await new Promise((resolve) => {
+			setTimeout(async () => {
+				const remainingTime = maximumWaitTime - (Date.now() - start);
+				return resolve(
+					await fetchValidTokenSet({
+						oktaId,
+						email,
+						openIdClient,
+						tokenSet: refreshedTokenSet,
+						refreshToken: newRefreshToken,
+						ip,
+						request_id,
+						maximumWaitTime: remainingTime,
+						waitTime,
+					}),
+				);
+			}, waitTime);
+		});
 
-	if (maximumWaitTime <= 0) {
-		// We've timed out waiting for the claim to be present
-		return null;
-	}
+	try {
+		const { legacy_identity_id } = tokenSet.claims();
+		if (legacy_identity_id) {
+			// The legacy_identity_id claim is present, we can return the token set
+			return tokenSet;
+		}
 
-	const start = Date.now();
-	const refreshedTokenSet = await openIdClient.refresh(refreshToken);
-	const newRefreshToken = refreshedTokenSet?.refresh_token;
-	if (!newRefreshToken) {
-		// There's no refresh token in this token set, so we can't continue
-		return null;
-	}
+		if (maximumWaitTime <= 0) {
+			// We've timed out waiting for the claim to be present
+			return null;
+		}
 
-	// Non-blocking wait for the next iteration
-	return new Promise((resolve) => {
-		setTimeout(async () => {
-			const remainingTime = maximumWaitTime - (Date.now() - start);
-			resolve(
-				await waitForLegacyIdentityId(
-					openIdClient,
-					refreshedTokenSet,
-					newRefreshToken,
-					remainingTime,
-					waitTime,
-				),
+		const start = Date.now();
+		// Attempt to fix the user profile.
+		const isOktaProfileFixed = await fixOktaProfile({
+			oktaId,
+			email,
+			ip,
+			request_id,
+		});
+		if (isOktaProfileFixed) {
+			const refreshedTokenSet = await openIdClient.refresh(refreshToken);
+			const newRefreshToken = refreshedTokenSet?.refresh_token;
+			if (!newRefreshToken) {
+				// There's no refresh token in this token set, so we can't continue
+				return null;
+			}
+			// Try again with the refreshed token set
+			return nonBlockingWait(
+				start,
+				maximumWaitTime,
+				waitTime,
+				refreshedTokenSet,
+				newRefreshToken,
 			);
-		}, waitTime);
-	});
+		}
+
+		// Try again but using the existing refresh token
+		return nonBlockingWait(
```

Comment:
I'm finding it hard to follow this with all the early returns. In what case does execution reach this ?
If the token already has an id then it returns the token so it doesn't get here
if we exceeded the wait time it returns null and we don't get here
if the profile is fixed then do the `nonBlockingWait` on line 170 with the refreshed tokenset 

so is this for the case where the token didn't have the id but and we had an error trying to fetch the id from idapi so we hope that it was a temporary issue and it would work next time ?

It would be good to have a flowchart or something like that somewhere to make sure we don't forget how this works

Reply:
Excellent idea, I've added a flowchart to the PR description!